Poly2TriMeshGenerator

roystgnr opened the pull request:
## Reason

This should be enough to close #20192, IMHO.  Getting Rocstar's capability may obsolete this eventually, but the timeline on that was out of our hands.

## Design

PolyLineMeshGenerator gives us the ability to specify arbitrary boundary loops, then Poly2TriMeshGenerator lets us triangulate boundaries specified by those loops (or by the boundary of some input 2D Mesh), with optional holes (that can either be left unfilled or can be stitched to the mesh defining them) defined by other input meshes, with various options for mesh and/or boundary refinement and for mesh smoothing.

## Impact

This adds two new mesh generators, along with tests and documentation.

Along the way I added input file support for `vector<bool>` parameters, and fixed a typo in the doc I was using as an example for how to write MOOSE docs.

roystgnr committed:
Added PolyLineMeshGenerator

Refs #20192

roystgnr committed:
Add PolyLineMeshGenerator regression tests

roystgnr committed:
Regression tests for refined polyline mesh gen

roystgnr committed:
Added Poly2TriMeshGenerator

roystgnr committed:
Add unit test for Poly2TriMeshGenerator

roystgnr committed:
Support holes in Poly2TriMeshGenerator

roystgnr committed:
Unit test Poly2TriMeshGenerator with holes

roystgnr committed:
Support vector<bool> parameters in hit

roystgnr committed:
Support vector<bool> parameters in parser

roystgnr committed:
Optional stitching of hole meshes to Poly2Tri

roystgnr committed:
Unit test for Poly2Tri triangulation stitching

roystgnr committed:
Add test of nested Poly2Tri trangulations

roystgnr committed:
Image of Poly2TriMeshGenerator output for docs

roystgnr committed:
Fix class description typo

roystgnr committed:
Add class descriptions for new mesh generators

roystgnr committed:
Poly2TriMeshGenerator Laplacian smoothing option

We need something better here but this is what was already in libMesh
and it ought to be good enough for now.

roystgnr committed:
Document PolyLineMeshGenerator

roystgnr committed:
Another Poly2Tri mesh picture for docs

roystgnr committed:
Document Poly2TriMeshGenerator

roystgnr committed:
Add regression test for Poly2Tri smoothing option

roystgnr committed:
Remove trailing whitespace

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1078685/) on 62c30fb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21303/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21303/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format d8857424f6409baa438ce06b4d4a9fa31f21a7b8`

roystgnr committed:
clang-format changes

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1078717/) on f0db335 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21303/site)

This comment will be updated on new commits.

GiudGiud commented:
Definitely needs a newsletter entry!

loganharbour commented on framework/include/meshgenerators/Poly2TriMeshGenerator.h:
```diff
@@ -0,0 +1,54 @@
+
+  std::unique_ptr<MeshBase> generate() override;
+
+protected:
```
Mind adding some doxygen comments in here? I also think that most of this can be const

roystgnr replied:
Done

loganharbour commented on framework/include/meshgenerators/PolyLineMeshGenerator.h:
```diff
@@ -0,0 +1,35 @@
+
+  std::unique_ptr<MeshBase> generate() override;
+
+protected:
```
Same here on doxygen comments and const. In general in moose we also try to avoid abbreviations when possible, i.e., `nebp` is wat?

roystgnr replied:
Done

loganharbour commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,241 @@
+                        false,
+                        "Whether to do Laplacian mesh smoothing on the generated triangles.");
+  params.addParam<std::vector<MeshGeneratorName>>(
+      "holes", std::vector<MeshGeneratorName>(), "The MeshGenerators that define mesh holes.");
```
Ugh I hate that this doesn't work

```suggestion
      "holes", {}, "The MeshGenerators that define mesh holes.");
```

roystgnr replied:
I tried removing the parameter entirely too (on the theory that that addParam API overload leaves values uninitialized but an "uninitialized" vector is the same as a default-initialized one), but that gives a runtime error.

Perhaps we could change the second `addParam` parameter from a `const S &` to a `T` (passing by value, then we move instead of copy later to keep the same efficiency)?  But that'd be quite an intrusive change I'd want to test in its own PR.

loganharbour replied:
Yeah - no worries. Just wanted to comment on my frustrations.

loganharbour commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,241 @@
+  // We put the boundary mesh in a local pointer
+  std::unique_ptr<UnstructuredMesh> mesh = dynamic_pointer_cast<UnstructuredMesh>(_bdy_ptr);
+
+  if (!_stitch_holes.empty() && _stitch_holes.size() != _hole_ptrs.size())
+    paramError("stitch_holes", "Need one stitch_holes entry per hole, if specified.");
+
+  if (!_interpolate_holes.empty() && _interpolate_holes.size() != _hole_ptrs.size())
+    paramError("interpolate_holes", "Need one interpolate_holes entry per hole, if specified.");
```
You should be able to check both of these in the constructor

roystgnr replied:
Moved

loganharbour reviewed (changes requested):
Moving @GiudGiud's comment to a request: Add a newsletter entry

roystgnr committed:
Only run Poly2Tri tests on ReplicatedMesh for now

We'll need to wait for a libMesh update with a fix there before we have
DistributedMesh triangulation properly working.

roystgnr committed:
Add new mesh generators to newsletter

roystgnr committed:
Mark const members as const

We can't do this with our unique_ptr input mesh because of the move
assignent / dynamic recasting that happens later, but for everything
else we set it from the input file and don't touch it again.

roystgnr committed:
_nebp -> _num_edges_between_points

Add a bit more readability, at Logan's suggestion

roystgnr committed:
Move paramError tests to constructors

roystgnr committed:
Better error reporting when smoothing

roystgnr committed:
Clarify some parameter descriptions

roystgnr committed:
Comment new mesh generator member variables

These mostly map 1-to-1 to parameters, but this way we get the
descriptions into Doxygen too.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1079641/) on 8e20c15 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21303/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21303/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ae3ac9a56db32885dbd184300c7417f50c0da3ca`

roystgnr committed:
Refactor new mesh generator tests files

Also link to the triangulator issue

roystgnr commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,241 @@
+  if (_smooth_tri)
+    for (auto elem : mesh->element_ptr_range())
+    {
+      libmesh_assert(elem->type() == TRI3);
```
Done, if you like the text.

roystgnr commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,241 @@
+      auto cross_prod = (elem->point(1) - elem->point(0)).cross(elem->point(2) - elem->point(0));
+
+      if (cross_prod(2) <= 0)
+        mooseError("Found inverted element in triangulation.  Disable Laplacian smoothing?");
```
I tried to expand it a bit, but could do more if you like.

loganharbour replied:
Looks good.

roystgnr commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,241 @@
+      hole_mesh.comm().max(new_hole_bcid);
+
+      ++new_hole_bcid;
+      const boundary_id_type inner_bcid = new_hole_bcid + 1;
```
Hmm... no, it's not, is it?  That's what I was trying for but I didn't think about conflicts with BCIDs due to other holes.  I'll fix it.

loganharbour replied:
I do something like this in one of the other generators. I wouldn't object to a general utility for this

GiudGiud replied:
yeah seems like we want this for block IDs too, not just boundaries @shikhar413  was running into this

roystgnr commented on framework/src/meshgenerators/PolyLineMeshGenerator.C:
```diff
@@ -0,0 +1,108 @@
+    paramError("points", "At least 2 points are needed to define a polyline");
+
+  if (_loop && _points.size() < 3)
+    paramError("points", "At least 3 points are needed to define a polygon");
```
Done

roystgnr commented:
This still needs a little work (I've done the BoundaryName thing, but I'll wait until I have any potential boundary id conflicts sorted tomorrow so I can push both commits together and only trigger one new CI sweep instead of two), but it's probably in good enough shape now to ping @eshemon in case she and her team might want a look at the interface design before it's merged.

roystgnr committed:
Use BoundaryName for PolyLineMeshGenerator bdys

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1079937/) on c17e713 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ae3ac9</th>
      <th colspan=4>#21303 c17e71</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.42%</td>
      <td>83.44%</td>
      <td>+0.02%</td>
      <td>95.03%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>77388</td>
      <td>77540</td>
      <td>+152</td>
      <td>153</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15378</td>
      <td>15387</td>
      <td>+9</td>
      <td>8</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21303/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21303/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21303/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,242 @@
+  params.addRequiredParam<MeshGeneratorName>(
+      "boundary", "The MeshGenerator that defines the mesh outer boundary.");
+  params.addParam<unsigned int>(
+      "interpolate_boundary", 0, "How many more nodes to add in each outer boundary segment.");
```
do you prefer interpolate over refine?
why not something explicit like num_nodes_per_segment ?

roystgnr replied:
I went with `add_nodes_per_boundary_segment` in the end; I figured `num` was misleading since this number isn't counting any part the nodes already at the beginning and end of each segment.

I think I've resolved all your issues now; would you mind double-checking and hitting "Resolve conversation" where ever I'm correct about that?  As a reviewer I sometimes find that an author's evaluation of "I've resolved this" doesn't match *my* evaluation, so I hate to be presumptuous/rude in the other direction.

GiudGiud commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,242 @@
+                                     "Whether to allow automatically refining each hole boundary.");
+  params.addParam<Real>(
+      "desired_area", std::numeric_limits<Real>::max(), "Desired (maximum) triangle area.");
+  params.addParam<std::string>("desired_area_func",
+                               std::string(),
```
```suggestion
  params.addParam<FunctionName>("desired_area_func",
                               std::string(),
```
does this work

GiudGiud replied:
oh this function is a parsed expression not a moose function?

roystgnr replied:
Correct; it goes into a libMesh API expecting an fparser friendly string.

GiudGiud commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,242 @@
+    meshed_holes.emplace_back(*_hole_ptrs[hole_i]->get());
+    if (hole_i < _refine_holes.size())
+      meshed_holes.back().set_refine_boundary_allowed(_refine_holes[hole_i]);
+    if (_interpolate_holes.size() > hole_i)
+      libmesh_not_implemented();
```
so "interpolation" or having more than one node per segment in holes is not implemented?

roystgnr replied:
Yeah; it turned out to be less trivial and more redundant (it's easy enough to refine the input defining the hole) than I anticipated.  I'll remove the parameter.

GiudGiud commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,242 @@
+            }
+        }
+      }
+      mooseAssert(found_hole_sides == np, "Failed to find full outer boundary of meshed hole");
```
how fool proof is this? Do we expect users to hit this in difficult cases?

GiudGiud replied:
in fact what if the user came in with a mesh that had a different number of outer points than what we used to mesh the hole outside?

roystgnr replied:
If I expected difficult cases to trigger it I'd have made it a mooseError; this should only trigger if we have an internal bug ... which is thinkable enough that I didn't want want to omit the assert entirely.

The hole is purely defined by the hole mesh ... but I suppose we could end up with an inconsistency before stitching if a user requested a hole be stitched but didn't disallow the hole from being automatically refined.  I'll add an error if that's ever inconsistent.

GiudGiud replied:
ok and what about if the hole has a hole?
or what if it s a 2D mesh?

roystgnr replied:
Holes with holes are supported (MeshedHole only finds an outer boundary, so then the inner boundary doesn't get the temporary stitching BCID assigned).  Holes defined by 2D meshes are supported - that's where the stitching option becomes useful.

GiudGiud commented on test/tests/meshgenerators/poly2tri_mesh_generator/poly2tri_mesh_generator.i:
```diff
@@ -0,0 +1,23 @@
+    desired_area = 0.05
+  []
+
+  parallel_type = replicated  # libMesh bug workaround
```
well will you really support distributed meshes?
seems like not to me even if poly2tri is/were distributed

we should have a check on this. use mesh.is_replicated or something like this

roystgnr replied:
Distributed meshes?  No.  `DistributedMesh`es?  Yes.  I threw a MeshSerializer in there, which ought to be fine for most "2D that won't get big until it's extruded to 3D" use cases.  I'm just not sure where or how it regressed; the poly2tri unit tests in libMesh ought to be running both replicated and distributed in CI.

GiudGiud replied:
so this needs to stay?
can you please link to the issue there so I can read about it or if it s not fixed soon we can just keep track of it in the future

roystgnr replied:
The libMesh side of the fixes is going in via https://github.com/libMesh/libmesh/pull/3324

That plus the MOOSE side of the fixes is working for me, so I'm also removing all the `parallel_type = replicated` lines there.  I'll hold off on pushing and triggering CI here until we've got the next libMesh submodule update done too, so until then I'm mirroring to a `mesh_generators_distributed` branch.

roystgnr replied:
> That plus the MOOSE side of the fixes ~~is working for me~~

I should never type "is working for me" about new DistributedMesh code until I've finished a sweep over processor counts.  Let's say "is getting a tiny bit closer for me", at least.

GiudGiud reviewed (changes requested):
so much capability offered here, real impressive

eshemon commented:
Thanks Roy. @aaronoaks Please review. I am also sending him the link in case the tagging doesn't work.

roystgnr committed:
Consider *all* hole meshes when finding temp bcids

roystgnr committed:
Rename param to add_nodes_per_boundary_segment

roystgnr committed:
Remove interpolate_holes param

This is a bit redundant (if we have a polyline hole we can easily
interpolate while generating it, and if we have a 2D mesh hole we
probably don't want to interpolate it and if we do we can refine it) and
the implementation was going to be a bit more hassle than I thought.

roystgnr committed:
paramError if we allow refining stitched hole bdys

This should catch any potential mismatch when stitching.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1081191/) on bd39949 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21303/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21303/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ad2d41eab153c0d864001ea684edc488368f62ca`

GiudGiud commented:
seems you are missing some headers for the non-unity build? MooseMeshUtils.h probably

let's remove the clang commits, squashing with the last one is fine, even better is to fixup the appropriate ones but let's not be picky

roystgnr committed:
Add MooseMeshUtils.h include

We were getting this indirectly in Unity builds but we need it directly
for non-Unity.

roystgnr commented:
> seems you are missing some headers for the non-unity build? MooseMeshUtils.h probably

That must have been it.  Thanks!

> squashing with the last one is fine

I strongly disagree.  Mixing automatically-generated changes with unrelated manual changes is adding noise to signal.  We don't even like libMesh re-bootstrap commits to be mixed with the changes they're generated from, and that's when the automatic changes are related to the manual changes and separated in completely distinct files.  Commit logs of "This is a meaningful change", "that is a whitespace change" convey useful information that is lost if they're squashed and the changes don't completely overlap.

> even better is to fixup the appropriate ones but let's not be picky

Well, it's work for dubious benefit rather than work for negative benefit, at least, so I took a crack at it.  `git filter-branch --tree-filter 'git clang-format' -- origin/devel..HEAD` followed by a `git rebase -i origin/devel` to get rid of the empty commits (and skimming `git diff origin/devel` to try to reassure myself that the edits in the rebased branch didn't extend outside my intended changes...) was easy enough, so I'm pushing the result.  Although, for some reason (different clang-format versions? different options?) it seems to only have emptied two out of the three clang-format commits.  I'd actually be interested in figuring out why I was unable to fix that automatically, albeit still strongly disinterested in trying to do so manually.

loganharbour commented:
> I strongly disagree. Mixing automatically-generated changes with unrelated manual changes is adding noise to signal.

I agree with this disagreement.

roystgnr commented:
Thanks for the extra set of eyes!

> When using an existing 2D mesh for the outer boundary, does the desired boundary name/id need to be specified? or is it able to discover the appropriate boundary?

If the 2D mesh has an outer boundary but also a number of holes, we autodetect the outer boundary and ignore the holes.  If the 2D mesh has multiple outer boundaries (it's not a connected domain), we pick one basically at random.  Not the best behavior, I admit, but easy to work around (triangulate the connected parts of the domain separately, then stitch) and it'd be hard to fix.

> Does the triangulated region have a predictable subdomain name or some other way to refer to the region later for assigning materials and such? If not can we add an option?

No, it doesn't, and it really should, shouldn't it?  I'd thought of that as redundant because we can pass the output from here to a later MeshGenerator to change subdomain ids, but that's only easy in the case where no hole meshes are stitched in; in the case where hole meshes are stitched in but we want the newly-triangulated elements to get a different subdomain ID, we ought to be assigning that ourselves here.  I'll add an option for that now, thanks.

GiudGiud commented on framework/include/meshgenerators/Poly2TriMeshGenerator.h:
```diff
@@ -0,0 +1,61 @@
+  std::unique_ptr<MeshBase> generate() override;
+
+protected:
+  /// Input mesh defining the boundary to triangulate
```
```suggestion
  /// Input mesh defining the boundary to triangulate within
```

GiudGiud commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,245 @@
+  params.addParam<std::vector<bool>>("refine_holes",
+                                     std::vector<bool>(),
+                                     "Whether to allow automatically refining each hole boundary.");
+  params.addParam<Real>(
+      "desired_area", std::numeric_limits<Real>::max(), "Desired (maximum) triangle area.");
```
make range checked

GiudGiud commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,245 @@
+                                     "Whether to allow automatically refining each hole boundary.");
+  params.addParam<Real>(
+      "desired_area", std::numeric_limits<Real>::max(), "Desired (maximum) triangle area.");
+  params.addParam<std::string>("desired_area_func",
```
what happens if function is negative over part of the domain?

GiudGiud commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,245 @@
+  poly2tri.set_refine_boundary_allowed(_refine_bdy);
+
+  poly2tri.desired_area() = _desired_area;
+  poly2tri.minimum_angle() = 0; // Not yet supported
```
any strong reason we dont support it?

roystgnr replied:
Because it'll take a long time to support.  It's a feature we can simply turn on with the Triangle interface, but Poly2Tri doesn't support refinement *at all* and I had to write refinement code from scratch.  Just handling all the corner cases from area-based refinement took quite a while.

GiudGiud commented on framework/src/meshgenerators/Poly2TriMeshGenerator.C:
```diff
@@ -0,0 +1,245 @@
+  // options, but even those might have failure cases.  Better to
+  // always do an extra loop here than to ever let users try to run on
+  // a degenerate mesh.
+  if (_smooth_tri)
+    for (auto elem : mesh->element_ptr_range())
+    {
+      mooseAssert(elem->type() == TRI3, "Unexpected non-Tri3 found in triangulation");
+      auto cross_prod = (elem->point(1) - elem->point(0)).cross(elem->point(2) - elem->point(0));
+
+      if (cross_prod(2) <= 0)
+        mooseError("Inverted element found in triangulation.\n"
+                   "Laplacian smoothing can create these at reentrant corners; disable it?");
+    }
```
should you add this to the SmoothMeshGenerator we already have then

roystgnr replied:
I didn't know that existed, but sure.  Obviously not in this PR, though.

GiudGiud commented on framework/src/meshgenerators/PolyLineMeshGenerator.C:
```diff
@@ -0,0 +1,112 @@
+    auto elem = Elem::build(EDGE2);
+    elem->set_node(0) = mesh.node_ptr(i);
+    elem->set_node(1) = mesh.node_ptr(ip1);
+    elem->set_id() = i;
```
this just makes it cheaper to add right

GiudGiud reviewed (commented):
Poly2Tri does sound real good.
I m worried it could be confused when we support Polygonal meshes as a converter from Polygon to Triangles. 
PolyLine2Tri ?

WilkAndy commented:
> Poly2Tri does sound real good. I m worried it could be confused when we support Polygonal meshes as a converter from Polygon to Triangles. PolyLine2Tri ?

How about SwissCheese2D ?

roystgnr commented:
> How about SwissCheese2D ?

Gonna say no to that one.  `PolyLine2Tri` isn't bad, but if we're not reflecting the underlying implementation then we might as well get away from `2tri` entirely.

How about `TriangulatorMeshGenerator`?  In hindsight this could be adapted pretty easily to use any other subclass of `libMesh::TriangulatorInterface`, even if licensing issues mean we're restricted to `poly2tri` for the near future.

GiudGiud commented:
TriangulatorMeshGenerator is less precise than specifying PolyLine.

In general I m worried with users finding something with the name that sort of matches what they want to do, but we have stronger limitations.
Do you support 1D sidesets of a 2D mesh for the polyline for example?

roystgnr commented:
> Do you support 1D sidesets of a 2D mesh for the polyline for example?

Not sidesets, but outer boundaries, yes.  We've already got some internal test coverage of that (it's one component of how the refinement works, when it has to make multiple sweeps), but I'll add an explicit test case too.  I just need to figure out one that is robust enough to put in an exodiff; simply throwing in a GeneratedMeshGenerator for the input is likely to give us a bunch of rectangular interior point arrangements, and then the diagonal selection on those might change at the whim of your compiler/CPU/etc.

Given that, does `TriangulatorMeshGenerator` make more sense?

GiudGiud commented:
Here's a few examples I tried to run (knowing they would fail but hoping they would be caught nicely).
1) off-plane, not 2D
```
  [./outer_bdy]
    type = PolyLineMeshGenerator
    points = '-1.0 0.0 0.0
              0.0 -1.0 0.01
              1.0 0.0 0.0
              0.0 2.0 0.0'
    loop = true
  []
  [./triang]
    type = Poly2TriMeshGenerator
    boundary = 'outer_bdy'
    add_nodes_per_boundary_segment = 4
    refine_boundary = false
    desired_area = 0.05
  []
```
in devel:
```
Assertion `p(2) == 0' failed.
p(2) = 0.0025000000000000001
0 = 0
```
in opt:
seg fault

2) random polyline, not even a loop
```
  [./outer_bdy]
    type = PolyLineMeshGenerator
    points = '-1.0 0.0 0.0
              0.0 -1.0 0.0'
  []
    [./triang]
    type = Poly2TriMeshGenerator
    boundary = 'outer_bdy'
    add_nodes_per_boundary_segment = 4
    refine_boundary = false
    desired_area = 0.05
  []
  ```
  in both opt and devel: "error not implemented" from libmesh, which really doesnt tell the user what's going on
  
3) random open stitched polyline. Gave me a mesh, but like why!!
```
[line1]
      type = PolyLineMeshGenerator
      points = '0.0 0.0 0.0
                1.0 0.0 0.0'
  []
  [line2]
    type = PolyLineMeshGenerator
    points = '1.0 0.0 0.0
              0.0 1.0 0.0'
  []
  [triangle_p1]
    type = StitchedMeshGenerator
    inputs = "line1 line2"
    stitch_boundaries_pairs = '2 1'
  []
  [./triang]
    type = Poly2TriMeshGenerator
    boundary = 'triangle_p1'
    add_nodes_per_boundary_segment = 4
    refine_boundary = false
    desired_area = 0.05
  []
```

Until you are 100% confident you can handle all the garbage a user can throw at it, I would not call it a TriangulatorMeshGenerator because that will attract a lot of attention

roystgnr commented:
> random polyline, not even a loop

I'm not sure "it fails to triangulate a non-closed polyline" is the best argument in *favor* of the name `PolyLine2Tri`.  Do you have a new suggested name?

> ```Assertion `p(2) == 0' failed.```

That can be tested for in opt; it'll just be more expensive to do so.  I'll add the tests.

> Until you are 100% confident you can handle all the garbage a user can throw at it

This will never happen; users are too creative with garbage.

Our remaining options are "never create a triangulator unless it gives a proper error message to every possible expected-to-fail user input *on the first try*", "never call a triangulator a triangulator", or "make everybody handle a name change to triangulator later when the error messages are perfect".  Do any of those sound reasonable?

What percentage of MOOSE (or any code!) would you say meets an analogous standard?  We just discussed up above that SmoothMeshGenerator can silently make a mesh inverted, not merely less smooth.  Should we rename it?

> I would not call it a TriangulatorMeshGenerator because that will attract a lot of attention

`Triangulator2DMeshGenerator`, perhaps?  The lack of 3D surface (and arguably volume, since the phrase "tetrahedralator" has never caught on) support is all that `Triangulator` hints at to me that we're not going to be able to provide within this framework.  Other than that, what attention are you worried about?  Do you really think someone is going to try to triangulate a non-closed boundary and be disappointed?

GiudGiud commented:
ClosedPolyLine2TriMeshGenerator ?

Calling something limited by the very general name, implying it can do something general, is indeed all over the framework. I've been steadily renaming these objects. It's very unrewarding work. But every time I do analysis and think I've found the object I need in the list (because of the name) only to have to dig in the source to understand why it doesnt work is also not pleasant. 

I havent tested this for closing up a 1D sideset (instead of a polyline) because it takes quite a few mesh generators to get it right. If you have an example I would be down to try it

GiudGiud commented:
Users are not going to know that there is an issue with the mesh/boundary they are providing. That it's not closed, that it's not in the right plane etc. Users do that all the friggin time. 

btw while I m being annoying, also hitting an unimplemented error with that 
```
  [./outer_bdy]
    type = PolyLineMeshGenerator
    points = '0.0 0.0 0
              0.0 0.0 1
              1.0 0.0 1.
              1.0 0.0 0'
    loop = true
  []
```

it's 2D just not the right plane

roystgnr commented:
If I have the two line segments in your test, is it obvious that the one correct way to connect them is via adding an extra segment, and is the difficulty of adding that extra segment before invoking the triangulator so extreme that it no longer deserves to be called a triangulator?

> implying it can do something general

It can do something general.  There is no 2D domain that can be triangulated that this (in conjunction with CombinerGenerator to handle non-connected domains and PolyLineMeshGenerator to define boundaries) cannot turn into a triangle mesh.  How much more general do we need before something deserves to be called a 2D triangulator?  It's not enough to triangulate any domain you define, it also has to handle incompletely defined domains first?  That sounds like a nice feature to add someday to handhold users, but it's not at all the definition of a triangulator.  Take a look at the [face.poly example](https://www.cs.cmu.edu/~quake/face.poly) in the award-winning Triangle.  That final segment closing the outer boundary isn't optional.  Look at [the API in Matlab](https://www.mathworks.com/help/matlab/ref/delaunaytriangulation.html) - if you want a constrained edge, it has to be manually specified ... and they won't even delete the domain-external triangles from the Delaunay triangulation for you!  From [SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.Delaunay.html) or [Octave](https://docs.octave.org/v4.4.0/Delaunay-Triangulation.html) or [the utility in their underlying qhull](http://www.qhull.org/html/qdelaun.htm) you can't even constrain edges, you just get the Delaunay triangulation of a point set convex hull.  (you also can't automatically generate interior points from any of the most popular triangulators save Triangle!).  Going down the list of Delaunay triangulators the first thing I find which has an option to close a polyline is CGAL, and even there it's hard to argue that "but you have to write C++ code and change a default option" counts as less difficult than "you have to change a default option in an input file".

Triangulating a non-closed boundary isn't even always defined.  If you have a single open-but-connected polyline, maybe "connect the first and last points" makes sense as an option (which is why this PR *already has it*, just in PolyLineMeshGenerator), but with as few as 4 disconnected segments it's easy to come up with a configuration that can be connected into multiple valid-but-differing domains.  Trying to triangulate incompletely defined domains isn't just a feature that this triangulator and Triangle and QHull etc. all lack, it's occasionally not even a well-defined problem.

In the end, if this isn't a triangulator, we need to fix up more than just this PR (and send crank mail to Shewchuk et. al.).  I currently have this marked as closing #20192.  Should I remove that, mark #20192 as "wontfix", and start writing up my "we tried to add a triangulator but only triangulating closed boundaries didn't count" report?

roystgnr commented:
> Users are not going to know that there is an issue with the mesh/boundary they are providing. That it's not closed, that it's not in the right plane etc. Users do that all the friggin time.

I agree!  And "we should detect and add an error message for that" is a great bug report!  But to say we can't call it a triangulator until then is just bikeshedding combined with ignorance of the history of the word "triangulator".

> it's 2D just not the right plane

Also a good criticism.  "XYTriangulatorMeshGenerator", then?

GiudGiud commented:
You're getting fixated on these open boundaries like it's the point I'm making but it's really not. We don't need to be able to triangulate these, we need a proper error message. Not segfault or "not implemented". There should some resiliency to user inputs. I'm sure there are plenty of famous packages that dont do things well. I've never hit a seg fault from Cubit or StarCCM meshing. 

To add to the list of common user errors, doubling a node also errors on "not implemented":
```
  [./outer_bdy]
    type = PolyLineMeshGenerator
    points = '0.0 0.0 0
              0.0 1.0 0
              0.0 1.0 0
              1.0 1.0 0
              1.0 0.0 0'
    loop = true
  []
```

And I ll stick that this isnt to be named a triangulator until it can do what many (not all) users will expect, which is triangulate 1D sidesets from 2D meshes. @RocksSavage is developing a mesh generator that will help me check that, but I already know looking at the input parameters that it is not geared for that. We'd want two things:
- input be the 2D mesh instead of the polyline
- have a sideset parameter instead of taking the whole mesh for the triangulation
We can roll with XYPoly2Tri temporarily.

roystgnr commented:
> We don't need to be able to triangulate these, we need a proper error message. Not segfault or "not implemented".

As I said, I agree!  What I strongly disagree with is giving a class a temporary name until every possible error has been covered with a message (and then what, after that? leave the same class with two names? force users to change from the old name to the new?).  Is this an actual current MOOSE policy?  Isn't either way user-unfriendly?

> I ll stick that this isnt to be named blankly a triangulator until it can do what many (not all) users will expect, which is triangulate 1D sidesets from 2D meshes.

Show me a triangulator which meets your definition.  I did the converse for you, repeatedly.  I gave you a list of the most famous triangulators online, only one of which did what you were asking for then, and *not a single one of which does what you demand here*; is that irrelevant?  Are you reading my responses?  Repeating

> input be the 2D mesh instead of the polyline

Seems to ignore what I already said,

> > Do you support 1D sidesets of a 2D mesh for the polyline for example?

> Not sidesets, but outer boundaries, yes.

I've got commits with a test case like this ready (retriangulating a circle), even, I'm just waiting to see what name we settle on first so I can add that change and push all together.

roystgnr commented:
I'll drop this for the weekend, though.  I've got more than enough other code to work on.  Sorry to get heated.

GiudGiud commented:
> > We don't need to be able to triangulate these, we need a proper error message. Not segfault or "not implemented".
> 
> As I said, I agree! What I strongly disagree with is giving a class a temporary name until every possible error has been covered with a message (and then what, after that? leave the same class with two names? force users to change from the old name to the new?). Is this an actual current MOOSE policy? Isn't either way user-unfriendly?

It's not that user-unfriendly to rename an object. There's a warning for it. 
The other option is not to merge this mesh generator because it is missing too much exception checking. I could user-break this 5 different ways and I havent even extracted 1D meshes from real meshes yet, like I was planning to do to test this. It needs to be resilient to go into production. I have never merged anything in MOOSE while knowing that it could seg fault beforehand, much less if it's going to attract as much attention as a Triangulator. 

> > I ll stick that this isnt to be named blankly a triangulator until it can do what many (not all) users will expect, which is triangulate 1D sidesets from 2D meshes.
> 
> Show me a triangulator which meets your definition. I did the converse for you, repeatedly. I gave you a list of the most famous triangulators online, only one of which did what you were asking for then, and _not a single one of which does what you demand here_; is that irrelevant? Are you reading my responses? Repeating

I would if i had been working on this for the last few months like you did. Pretty sure Cubit can mesh a sideset off of a mesh. With quads too. Doesnt even have to be coplanar if there's a volume defined under it. 

> I've got commits with a test case like this ready (retriangulating a circle), even, I'm just waiting to see what name we settle on first so I can add that change and push all together.

I've suggested a couple names that are acceptable in my mind, that do not mislead users about what they can / cannot do.
The 1D prefix come to think of it might not be very accurate/clear.
If you choose XYBoundaryMeshTriangulator, we will be able to add XYSidesetTriangulator fairly soon with Trent's new MG.

roystgnr commented:
> The other option is not to merge this mesh generator because it is missing too much exception checking.

That's a more reasonable option, short-term.

> I could user-break this 5 different ways

Do so, please?  5 more bug reports would be great.

> Pretty sure Cubit can

My fault here for not being clear enough.  I'm looking for something *calling itself* a "Triangulator", not just incorporating triangulators; I certainly wouldn't call this "CubitMeshGenerator" because I wouldn't dare claim it's anywhere near equivalent to Cubit.  Similarly, when MOOSE gets general triangulation capabilities, that's not going to be evidence that future triangulators had better pick a different name until they can handle nuclear physics too.

roystgnr commented:
> With quads too.

Just to make sure, you're not insinuating that we might have to wait until a code has [quad](https://en.wiktionary.org/wiki/quad-) support before calling it a [tri-](https://en.wiktionary.org/wiki/tri-#Etymology_1)angulator, right?

roystgnr commented:
So that's a "no", then.  What was the purpose of that line, then?  I thought I was either answering a rudely irrelevant "you're not as good as Cubit" snipe with less-rude humor, or I was checking on an actual feature demand.  Is "it has to create quads" really a priori *obviously* a less reasonable demand than "it has to be able to start from sidesets", when you're citing the same code to justify the latter and simultaneously bringing up the former?

aaronoaks commented:
@roystgnr I had another thought looking at `PolyLineMeshGenerator`, there is the option to add boundary names to the two ends of a non-looped line, could we add an option to add a boundary name to a looped polyline so that we can refer to the polyline later for adding boundary conditions? Like for the `poly2tri_nested.i` example case in the `Poly2TriMeshGenerator` docs, applying a BC on the `layer_4_bdy` outer boundary of the triangulated `triang_4` mesh.

permcody commented:
I've been asked to weigh in by multiple bystanders now. These kinds of debates are as old as the peer-review process and having been on both sides of these kinds of debates I can empathize with all of this including the more heated parts. You both make many good points but real software development deployed in real scenarios is just messy business. In general, I think it's good to name objects with the intent that they may fulfill even if they aren't there yet. Does that mean we should expect everything to work? Probably not on this planet. There may be reasonable ways to convey which parts are robust, which are a little jagged and those which are downright scary. Hell, this could just be done in the docs or class description. Also, We could open several new tickets for cases that should be worked in the future. 

The downside with giving it a really specific narrow use-case name is that it might just be missed by an analyst coming in wondering if an object exists. I really don't think there's a ton or risk with them thinking that it'll work for every situation. On the other hand, I've done user support long enough to know that users will make some very silly mistakes and a few honest ones and if we can provide better diagnostics, we really should. I would use errors more than asserts for "experimental" code. We can take the performance hit (premature optimization is the root of all evil). 

More than anything, I hate to hold back good chunks of new code until they are perfect because they likely _never_ will be. Let's do our best, document the shortcomings, and move on. When somebody hits issues, we'll come back. Guillaume thanks for your super thorough reviews, they are much appreciated. Roy, thanks for your tireless work/re-work to polish it to the best of your ability. Let's come together and get it done!

roystgnr commented:
> could we add an option to add a boundary name to a looped polyline so that we can refer to the polyline later for adding boundary conditions?

I'm afraid I'm not sure we can at that point; each edge only recognizes its two endpoints as boundaries.  The idea of an inner and outer boundary to a manifold is what the "shellface" boundary type was for, but I'm not sure anybody uses that (and I'm not sure libMesh supports that!) except for the TRI3SUBDIVISION use cases.

> Like for the poly2tri_nested.i example case in the Poly2TriMeshGenerator docs, applying a BC on the layer_4_bdy outer boundary of the triangulated triang_4 mesh.

At *this* point, though, I think we're in business.  We should already be adding boundary ids to each new triangulation (except of course I'm not even documenting that outside the code, am I; I'll fix that...), but we give users no way to name them and no control over boundary ids.  You could add an extra RenameBoundaryGenerator layer to each input mesh to avoid conflicts and then to the output mesh to put ids back where you want them ... but no.  I think I'd be happy with calling a one-extra-layer solution "modular" but two feels like a hack.  I'll add this.

roystgnr commented:
> If you're not willing to address the review then dismiss it and pick another reviewer.

Frankly this thought horrifies me.  No offense to Logan, he was helpful too, but you've caught several real issues both he and I missed and I definitely don't want things like that to get swept under the rug.  I apologize for being rude, and you should call me on it if I slip up again.

Can we postpone the name question for now, maybe just put it up for a vote in C510 at some point, but still continue with the bug hunting and feature requests?

Let me try to extract the issues list from all the mess...  I'll mark the ones that are fixed in my branch  I've got:

Things fixed in my local branch; I'll push again soon and make sure CI agrees:
* Range-checking applicable parameters
* Additional test cases
* * Re-triangulating a 2D mesh

Things that we agree need to be fixed here:
* Even in opt builds, even bad user inputs should end in a related error message, never a segfault or infinite loop or just "not implemented, line X", including:
* - [ ] Non-closed inputs
* - [ ] Non-xy-plane inputs
* - [ ] Duplicated nodes
* - [ ] Desired-area-functions returning non-positive values

Things we might as well fix along the way (the desired_area_func and some of those "not implemented" calls are going to take a libMesh update first, at which point I can slip in more at that level):
- [ ] Support for non-ReplicatedMesh (I'd call this a feature request except it was *supposed* to be already working, the existing support isn't even failing cleanly, and it'll be easier to fix the remaining errors than to add error messages)
- [ ] Detection of multiple-outer-boundary, disconnected-domain cases, so we can respond with error-messages instead of eenie-meenie-minie-moe.

Feature requests to do:
- [ ] Support for selecting a subdomain id + name for the newly-triangulated subdomain
- [ ] Support for manually setting boundary ids + names for the outer boundary and/or hole boundaries

Feature requests to fight about when to do:
- [ ] Support for using only a specified sideset of an input mesh as a boundary definition rather than the mesh outer boundary.
* Support for using the boundary of a block of an input mesh?  Is this redundant with the above, since users could just squeeze in a SideSetsAroundSubdomainGenerator?

Although I'll still die on my hill fighting against "you can't call something a triangulator without this feature", if I step out from that tunnel vision I find I don't have any quibble with "this feature would be a very good idea to have".  I think it'll be harder to add than the rest of the lists here so I'm not confident about getting it into this branch ASAP, but I'll at least take a crack at it and hope I'm wrong.

Anything I missed?

GiudGiud commented:
```
  [./outer_bdy]
    type = PolyLineMeshGenerator
    points = '0.0 5.0 0
              0.0040820584371332996 5.082246634058751 0
              0.032459620243681146 5.325083816716627 0
              0.10845166250253213 5.716992295567392 0
              0.2534598002445025 6.239392688983062 0
              0.4860975401805023 6.867539048506324 0
              0.8214214865058571 7.571700996856079 0
              1.270292671775893 8.318570459349798 0
              1.8388914764735926 9.07281734481298 0
              2.5284040419434395 9.798711272287777 0
              3.3348918737970106 10.461723979327065 0
              4.249349701100716 11.030029492997972 0
              5.257949840368312 11.475826306710855 0
              6.34246455900798 11.77641721268863 0
              7.48085148459144 11.914997367166663 0
              8.64798119723831 11.881118682984466 0
              9.816480986177442 11.670817674249239 0
              10.957664534587717 11.286413262214726 0
              12.042514171619938 10.735999624253228 0
              13.042680410090924 10.032675826481944 0
              13.931462842211307 9.193567754310441 0
              14.684737117299532 8.238707963216545 0
              15.28179465128425 7.18984497048066 0
              15.706064848393723 6.069254921488402 0
              15.945693837719826 4.8986254978579336 0
              15.993958888818534 3.698074673944052 0
              15.849503584717477 2.4853560154839194 0
              15.516385283989264 1.2752884116848242 0
              15.003933162778688 0.07943237161334082 0
              14.326421948647752 -1.0939816627783931 0
              13.50257309439841 -2.2398840505843345 0
              12.554901351896286 -3.355982097676404 0
              11.508930268937672 -4.4422011090221085 0
              10.392304845413262 -5.500000000000003 0
              9.233833278175616 -6.531617708612882 0
              8.062492263728672 -7.539309317136663 0
              6.906431622428712 -8.524629441970145 0
              5.792014010617905 -9.487815254835095 0
              4.7429241979941565 -10.427312877546857 0
              3.779379853229186 -11.339479486552223 0
              2.917472093059084 -12.218480087890283 0
              2.168659342148844 -13.05638351369949 0
              1.539432493205195 -13.843447740506003 0
              1.0311631500206897 -14.568571120910935 0
              0.6401401047702711 -15.21987446347863 0
              0.3577923846801229 -15.785369866255687 0
              0.1710904485560671 -16.253666398542343 0
              0.06311066477055718 -16.614660501719065 0
              0.013742292284473073 -16.86016046555071 0
              0.0005110287600811755 -16.984399401025453 0
              -0.0005110287600811849 -16.984399401025453 0
              -0.013742292284473158 -16.86016046555071 0
              -0.06311066477055745 -16.614660501719065 0
              -0.17109044855606656 -16.253666398542347 0
              -0.35779238468012375 -15.785369866255687 0
              -0.6401401047702698 -15.219874463478629 0
              -1.0311631500206915 -14.568571120910931 0
              -1.5394324932051928 -13.843447740506008 0
              -2.168659342148845 -13.056383513699487 0
              -2.9174720930590805 -12.218480087890283 0
              -3.7793798532291905 -11.339479486552221 0
              -4.742924197994154 -10.427312877546857 0
              -5.792014010617905 -9.487815254835093 0
              -6.9064316224287055 -8.524629441970147 0
              -8.062492263728675 -7.539309317136662 0
              -9.233833278175613 -6.531617708612886 0
              -10.39230484541327 -5.499999999999994 0
              -11.508930268937675 -4.442201109022106 0
              -12.554901351896286 -3.3559820976764048 0
              -13.502573094398405 -2.239884050584338 0
              -14.326421948647756 -1.0939816627783876 0
              -15.003933162778688 0.07943237161334449 0
              -15.516385283989264 1.2752884116848224 0
              -15.849503584717471 2.485356015483917 0
              -15.993958888818534 3.69807467394406 0
              -15.945693837719826 4.898625497857939 0
              -15.706064848393723 6.069254921488398 0
              -15.281794651284255 7.189844970480653 0
              -14.684737117299532 8.238707963216548 0
              -13.931462842211307 9.19356775431044 0
              -13.042680410090924 10.03267582648194 0
              -12.042514171619933 10.735999624253232 0
              -10.957664534587712 11.286413262214726 0
              -9.816480986177442 11.670817674249239 0
              -8.647981197238318 11.881118682984466 0
              -7.480851484591434 11.914997367166661 0
              -6.342464559007974 11.77641721268863 0
              -5.257949840368312 11.475826306710854 0
              -4.249349701100723 11.030029492997977 0
              -3.334891873797007 10.461723979327058 0
              -2.528404041943438 9.798711272287775 0
              -1.8388914764735933 9.072817344812986 0
              -1.2702926717758953 8.318570459349806 0
              -0.8214214865058549 7.571700996856076 0
              -0.4860975401805023 6.867539048506324 0
              -0.2534598002445028 6.239392688983065 0
              -0.10845166250253133 5.716992295567389 0
              -0.03245962024368097 5.325083816716626 0
              -0.004082058437133302 5.082246634058752 0
              0 5.0 0'
    loop = true
    show_info = true
  []
  [./triang]
    type = Poly2TriMeshGenerator
    boundary = 'outer_bdy'
    refine_boundary = false
  []
```
I'm sorry the disagreements became so stark as well. I should have been less upfront and more collaborative about this.

shikhar413 commented:
@roystgnr sorry to join the conversation so late into the review process, but I was testing out these new mesh generators and I ran into some issues with stiching holes that maybe you could take a look into? I have attached the input file that I am using but for some reason the mesh generator insists on cutting out the hole mesh in the problem, even if I set stitch_holes to true. I can't figure out why this is happening, but it could easily be a problem with how I've defined the input, thought I'd bring up the issue here as an additional validation problem for these MG's. One note is the created mesh relies on the mesh generators in the Reactor module to define the circular pin regions in the problem. Overall a very useful functionality though, thank you for bring this to fruition!

![image](https://user-images.githubusercontent.com/14334316/176661740-2e568684-3dda-4e50-bda7-3bebb0baf0b9.png)

[poly2tri_input.txt](https://github.com/idaholab/moose/files/9019180/poly2tri_input.txt)

roystgnr commented:
> so late into the review process

Not late at all, sadly!  I've got a bunch of issues addressed in a fork of this branch, plus a libMesh update that should *finally* be ready to merge there this afternoon, but there's still a ton more work to go, and even if there hadn't been, post-merge bug reports are always welcome, pre-merge are even better.

> it could easily be a problem with how I've defined the input

Looks like it to me.  From your attached poly2tri_input.txt, I see `stich_holes = 'true true'`, whereas what we'd need there is `stitch_holes = 'true true'`.

I don't understand why MOOSE isn't screaming at us about seeing an input file option that is never queried by its corresponding object, though.  Did MOOSE lose UFO detection in the move from GetPot to Hit?  Did MOOSE never enable input file UFO detection in the first place?  Being able to type `super_importnt_flag = 'nondefault'` and have the code silently execute with `super_important_flag = 'default'` anyway because of a typo is really scary!

GiudGiud commented:
--mesh-only does not have unused parameter detection.
There's higher priority items imo but if you feel strongly about this one you're welcome to weigh in when priorities are drafted

shikhar413 commented:
Thanks @roystgnr that change did the trick. Thank you for looking into this!

roystgnr commented:
> I started getting segfaults trying to set it to 33 or less.

@aaronoaks - thanks for posting these!  It was easy to reproduce these segfaults, extremely hard to diagnose them ... but in hindsight the bug was obvious once I saw it.

The fix is on the libMesh side, though, in https://github.com/libMesh/libmesh/pull/3339

I think I'm almost done with fixes there, at which point I'll get a submodule update going.

roystgnr commented:
Obsoleted by #21800
